native/src/main/java/io/ballerina/stdlib/data/csvdata/csv/CsvTraversal.java

```diff
@@ -0,0 +1,802 @@
+        HashSet<String> fields = new HashSet<>();
+        Type restType;
+        Deque<String> fieldNames = new ArrayDeque<>();
+        BArray rootCsvNode;
+        Type expectedArrayElementType;
+        Type sourceArrayElementType;
+        CsvConfig config;
+        String[] headers = null;
+        int arraySize = 0;
+
+        void reset() {
+            currentCsvNode = null;
+            currentField = null;
+            fieldHierarchy.clear();
+            updatedRecordFieldNames.clear();
+            headerFieldHierarchy.clear();
+            fields.clear();;
+            restType = null;
+            fieldNames.clear();
+            rootCsvNode = null;
+            expectedArrayElementType = null;
+            sourceArrayElementType = null;
+            config = null;
+            headers = null;
+            arraySize = 0;
+        }
+
+
+        void resetForUnionTypes() {
+            currentCsvNode = null;
+            currentField = null;
+            fieldHierarchy.clear();
+            updatedRecordFieldNames.clear();
+            headerFieldHierarchy.clear();
+            fields.clear();;
+            restType = null;
+            fieldNames.clear();
+            rootCsvNode = null;
+            expectedArrayElementType = null;
+            headers = null;
+            arraySize = 0;
+        }
+
+        CsvTree() {
+            reset();
+        }
+
+        @SuppressWarnings("unchecked")
+        public Object traverseCsv(BArray csv, CsvConfig config, Type type) {
+            this.config = config;
+            sourceArrayElementType = TypeUtils.getReferredType(getSourceElementTypeForTupleAndArrays(csv));
+            Type referredType = TypeUtils.getReferredType(type);
+            int sourceArraySize = (int) csv.getLength();
+            if (referredType.getTag() == TypeTags.INTERSECTION_TAG) {
+                Optional<Type> mutableType = getMutableType((IntersectionType) referredType);
+                if (mutableType.isPresent()) {
+                    return CsvCreator.constructReadOnlyValue(traverseCsv(csv, config, mutableType.get()));
+                }
+            }
+
+            if (referredType.getTag() != TypeTags.UNION_TAG) {
+                if (referredType.getTag() == TypeTags.ARRAY_TAG) {
+                    Type arrayElementType = TypeUtils.getReferredType(((ArrayType) referredType).getElementType());
+                    if (arrayElementType.getTag() == TypeTags.INTERSECTION_TAG) {
+                        Optional<Type> mutableType = getMutableType((IntersectionType) arrayElementType);
+                        if (mutableType.isPresent()) {
+                            return CsvCreator.constructReadOnlyValue(
+                                    traverseCsv(csv, config, TypeCreator.createArrayType(
+                                            mutableType.get()
+                                    ))
+                            );
+                        }
+                    }
+                }
```

Comment:
I feel like this is too deeply nested, shall we factor the logic to a separate function that returns an Optional (within it we can use early returns to avoid unnecessary nesting) and return here based on that